text/2019-01-08-pd-simulator.md

```diff
@@ -0,0 +1,150 @@
+the `eventRunner` will execute the `Run` function. The concrete definition is
+designed as follows:
+
+```golang
+type Event interface {
+    Run(tickCount int64) bool
+}
+```
+
+It runs the predefined events according to the `tickCount`. The return value
+indicates if the event is finished.
+
+### Basic process
+
+To run the simulator, we need to write the case which we want to test. The basic
+case usually consists of these parts:
+
+- Stores: It is related to stores' settings including number, status, capacity,
+  etc.
+- Regions: It is related to Regions' settings including number, distribution,
+  etc.
+- RegionSplitSize (optional): When the size of a Region exceeds it, the Region
+  will split.
+- RegionSplitKeys (optional): When the number of keys of a Region exceeds it,
+  the Region will split.
+- Events: It defines the event the case will run. If we don't need it, we can
+  initialize it with an empty list.
+- Checker: It checks if the case runs as expected.
+
+When we start the simulator, it will first create a driver, which will initilize
+the mocking TiKV cluster according to the specified case. After the PD is
+bootstrapped, it will start to run a timer. The driver will trigger each node to
+send the heartbeat to PD according to the tick count. In order to simulate a
+real TiKV cluster's behavior, for each tick, it will execute some Raft logic,
+such as leader election, Region split, on a shared `raftEngine` if needed and
+execute the events predefined in the case. Besides, after receiving the PD's
+response, it will convert this response to the task, and put this task into a
+task list of a node.
+For each tick, it will run these tasks to execute some Raft commands on the
+shared `raftEngine`.
+
+#### Verification of correctness
+
+To verify the correctness of the scheduling of PD, we need to define a checker.
+The checker will continue to check the condition we specified. If the condition
+is met, which means the scheduling result is as we expected, it will exit.
+Otherwise, it will keep running until its timeout expires and then raise an
+error.
+
+#### Evaluation of performance
+
+To evaluate the performance of PD under high concurrency requests, we can let
+the condition of the checker always return false, and we can use the metrics of
+PD to investigate the performance problem.
+
+## Drawbacks
+
+Since it's a new tool for PD, there are no drawbacks, but there are some points
+could be optimized:
+
+- The cases are written in `go`. It is necessary to recompile the code to run
+  the new case. It will be better to use the configuration file to write the
+  case directly.
+- It needs to add some metrics for the simulator to help locate problems.
+- It cannot simulate the RocksDB behavior, which needs to be improved in the
+  future.
+
+## Alternatives
+
+None
+
+## Unresolved questions
+
+The case is not easy to be constructed. We need to find a way to make it easier.
```

Comment:
```suggestion
The case is not easy to construct. We need to find a way to make it easier.
```

Reply:
- The box is easy to carry. (right)
- The box is easy to be carried. (wrong)
In "be+adj.+to do", we use the active voice to indicate the passive voice.